### DISCUSSION TITLE
Does EdgeDB work out of the box with AWS Aurora w/ PostgreSQL engine?

### DISCUSSION THREAD

**Discussion by ediskandarov-cur** [geldata/gel#3067] [Q&A] [answered]
I'm planning to use AWS for infrastructure.

Does EdgeDB work out of the box with AWS Aurora w/ PostgreSQL engine?

**Comment by fantix** [comment, marked as answer]
In short: yes!

For basic use cases and most advanced use cases covered by the [current server test suite](https://github.com/edgedb/edgedb/actions/runs/1372808650), it's positive they work fine on AWS RDS Aurora with PostgreSQL (12.8 tested, 13 should be fine too). Read replicas are not supported yet, even though you can access the Aurora read replicas directly fwiw. High availability of Aurora is supported now through the [Adaptive Backend HA](https://www.edgedb.com/docs/internals/backend_ha#adaptive-ha) feature, I'll create a PR later to support native RDS HA through AWS API.

**Reply by ediskandarov-cur** [reply]
Got it. Thanks! I'll give it a try.

**Reply by dancomanlive** [reply]
@emcpow2 Any updates on your progress?

**Reply by buzinas** [reply]
@fantix is it already possible to use read replicas with EdgeDB?

We're starting a multi-tenant / multi-language application and we're considering Aurora Global or Fly.io multi-region Postgres with read replicas, but I couldn't find any guides on how to set up EdgeDB to properly handle this use case.

I tried to set up a very simple Fly.io in 2 separate regions but couldn't find a way to force EdgeDB use one master region for writes.

Any thoughts here? Could you provide me some guidance and/or an update about read replicas?

**Reply by fantix** [reply]
Postgres read replicas are not natively leveraged by EdgeDB itself yet, it'll be some work but not a lot practically as we picture how to do it though. Before that happens, I don't anticipate anything that may stop you from designing EdgeDB read replicas from an application level, e.g. you could start two EdgeDB instances in each region, and make your application always use the one with the primary Postgres cluster for all the writes. If you are familiar with the EdgeDB [binary protocol](https://www.edgedb.com/docs/reference/protocol/index), you could also come up with a simple proxy server like pgPool, the "[capabilities](https://www.edgedb.com/docs/reference/protocol/messages#commanddatadescription)" flag of a query can be used to know if the query is read-only or not.

**Reply by seivan** [reply]
@fantix I mean, having multiple EdgeDB instances doesn't exactly the same thing as having multiple endpoints for your RDS instance for read queries, or does EdgeDB do something to mitigate expensive queries and connection pooling occurring to that read replica?

**Reply by jackfischer** [reply]
Hey @fantix, any update on native RDS HA on the AWS API?

**Reply by jackfischer** [reply]
Alternatively, would you expect the "EdgeDB read replicas from an application level" technique to work with later versions of edgedb that do writes for the query cache?

**Reply by fantix** [reply]
@seivan oh sorry I missed your question:

> does EdgeDB do something to mitigate expensive queries and connection pooling occurring to that read replica?

It doesn't do so at this moment, but it would be an interesting and feasible feature to add. Still, given the data consistency/synchronization limitations between replicas, designing a generic solution in EdgeDB to balance application tolerance and overall performance can be a bit complicated. Starting one EdgeDB server per Postgres replica would still be more controllable on the application end.

> any update on native RDS HA on the AWS API?

@jackfischer Negative, sorry it's not prioritized :(

> would you expect the "EdgeDB read replicas from an application level" technique to work with later versions of edgedb that do writes for the query cache?

Yes, that should work; we have all the logistics to synchronize persistent cache among different EdgeDB as "tenants" (as for the same Postgres cluster/replica fleet) done with underneath pg notifications and so on. However, if you have a lot of cached queries, say more than 1000 (`_MAX_QUERIES_CACHE`), the invalidation may race between tenants IF they have different query variety profiles, resulting in a relatively lower cache hit rate due to the fact that the cache invalidation is currently done in a per-tenant distributed/relaxed way to be fast and simple.

**Reply by jackfischer** [reply]
Understood on RDS HA support, especially with Cloud coming to market. That's great news on the query cache and so cool to hear how much thought has gone into it.
